### PR TITLE
fix #3745: throw better exceptions when a namespace is missing

### LINKED ISSUE #3745: createOrReplace on Secrets throws in case the resource already exists
### Describe the bug

Version of the `kubernetes-client`: 5.11.2

Performing 2 `createOrReplace` on the same Secret cause the second call to fail with this error:

```
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://192.168.64.3:8443/api/v1/secrets/test. Message: the server could not find the requested resource. Received status: Status(apiVersion=v1, code=404, details=StatusDetails(causes=[], group=null, kind=null, name=null, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=the server could not find the requested resource, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=NotFound, status=Failure, additionalProperties={}).
```

### Fabric8 Kubernetes Client version

next (development version)

### Steps to reproduce

Running a Main like this would reproduce the issue:

```java
public class Main {

    private static Secret getSecret() {
        return new SecretBuilder()
                .withNewMetadata()
                .withName("test")
                .withNamespace("default")
                .endMetadata()
                .build();
    }
    
    public static void main(String[] args) {
        Config config = new ConfigBuilder().withNamespace(null).build();
        KubernetesClient client = new DefaultKubernetesClient(config);

        client.secrets().createOrReplace(getSecret());
        client.secrets().createOrReplace(getSecret());
    }
    
}
```

### Expected behavior

The resource is seamlessly replaced.

### Runtime

minikube

### Kubernetes API Server version

1.22.3@latest

### Environment

macOS

### Fabric8 Kubernetes Client Logs

```shell
io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://192.168.64.3:8443/api/v1/secrets/test. Message: the server could not find the requested resource. Received status: Status(apiVersion=v1, code=404, details=StatusDetails(causes=[], group=null, kind=null, name=null, retryAfterSeconds=null, uid=null, additionalProperties={}), kind=Status, message=the server could not find the requested resource, metadata=ListMeta(_continue=null, remainingItemCount=null, resourceVersion=null, selfLink=null, additionalProperties={}), reason=NotFound, status=Failure, additionalProperties={}).
    at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure (OperationSupport.java:683)
    at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure (OperationSupport.java:662)
    at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode (OperationSupport.java:613)
    at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse (OperationSupport.java:556)
    at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse (OperationSupport.java:519)
    at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleGet (OperationSupport.java:488)
    at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleGet (OperationSupport.java:458)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.handleGet (BaseOperation.java:696)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.getMandatory (BaseOperation.java:182)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.require (BaseOperation.java:163)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.requireFromServer (HasMetadataOperation.java:120)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.replace (HasMetadataOperation.java:187)
    at io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation.replace (HasMetadataOperation.java:141)
    at io.fabric8.kubernetes.client.utils.CreateOrReplaceHelper.replace (CreateOrReplaceHelper.java:68)
    at io.fabric8.kubernetes.client.utils.CreateOrReplaceHelper.createOrReplace (CreateOrReplaceHelper.java:60)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace (BaseOperation.java:316)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace (BaseOperation.java:83)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace (BaseOperation.java:306)
    at io.fabric8.kubernetes.client.dsl.base.BaseOperation.createOrReplace (BaseOperation.java:83)
    at org.keycloak.Main.main (Main.java:27)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:254)
    at java.lang.Thread.run (Thread.java:829)
```

### Additional context

I suspect the issue sits in the `replace` method, but I haven't narrowed it down.

### PR DESCRIPTION
**Pull request by shawkins**
## Description
This differs from what we talked about in that it addresses #3745 without introducing a "default" namespace.  This will simply produce better error messages.

## Type of change

 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](../CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Review by rohanKanojia — APPROVED** [review]

**Comment by rohanKanojia** [comment]
test are failing :
```
[INFO] 
Error:  Failures: 
Error:    BaseOperationTest.testHttpRetryWithMoreFailuresThanRetries:300 As the last failure, the 3rd one, is not an IOException the message expected to contain: 'Internal Server Error'! ==> expected: <true> but was: <false>
Error:  Errors: 
Error:    BaseOperationTest.testHttpRetryWithLessFailuresThanRetries:316 » KubernetesClient
Error:    BaseOperationTest.testNoHttpRetryWithDefaultConfig:279 NullPointer
[INFO] 
Error:  Tests run: 122, Failures: 1, Errors: 2, Skipped: 0
```

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/BaseOperation.java [hunk 1]
```diff
@@ -291,31 +290,27 @@ public R load(String path) {
...
     T itemToCreateOrReplace = getItem();
+    Resource<T> resource;
     if (items.length > 1) {
       throw new IllegalArgumentException("Too many items to create.");
     } else if (items.length == 1) {
       itemToCreateOrReplace = items[0];
+      resource = withItem(itemToCreateOrReplace);
+    } else {
+      resource = this;
     }
```
Comment (manusa) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/internal/BaseOperation.java:301]: I find this sort of confusing, maybe replacing with the following might help:
```suggestion
    final T itemToCreateOrReplace;
    Resource<T> resource;
    if (items.length > 1) {
      throw new IllegalArgumentException("Too many items to create.");
    } else if (items.length == 1) {
      itemToCreateOrReplace = items[0];
      resource = withItem(itemToCreateOrReplace);
    } else {
      itemToCreateOrReplace = getItem();
      resource = this;
    }
```

**Comment by shawkins** [comment]
@rohanKanojia the test failures are unrelated to this change, those involve temporal/retry logic that is prone to failure on github.

**Review by manusa — APPROVED** [review]

**Comment by shawkins** [comment]
> @rohanKanojia the test failures are unrelated to this change, those involve temporal/retry logic that is prone to failure on github.

Arg I was only testing when I was logged into a kubernetes instance.  Without doing that I get the exception as there is no namespace specified on the config.  The tests have been updated.

**Comment by manusa** [comment]
I'll try this on my local machine

**Comment by manusa** [comment]
Apparently there's a problem with Maven 3.8.5, Windows and Sundr.io

**Comment by manusa** [comment]
After https://github.com/fabric8io/kubernetes-client/pull/3989, Windows build should work

**Pull request merged**